Standard SATB Barrier Implementation

RSalman opened the pull request:
These changes implement the the SATB access barriers, these have been adapted from the Realtime access barriers.

- Removed Double Barrier Code, Standard SATB does not make use of it

Adapted the following barrier from realtime for Standard SATB:

- forcedToFinalizableObject
- referenceGet
- referenceReprocess
- jniDeleteGlobalReference
- ~storeObjectToInternalVMSlotImpl~
- ~readObjectFromInternalVMSlotImpl~
- stringConstantEscaped / checkStringConstantsLive
checkClassLive

Signed-off-by: Salman Rana <salman.rana@ibm.com>

RSalman commented:
I know  barrier implementation for `checkClassLive `, `storeObjectToInternalVMSlotImpl` and `readObjectFromInternalVMSlotImpl` are excessive  and  not needed (most likely). However, I'm not comfortable removing them as I did full testing with them implemented. Implementing these has a performance hit, but I think that's acceptable since SATB still under dev (i.e JIT not enabled, requires OOL allocations).

amicic commented on runtime/gc_modron_standard/StandardAccessBarrier.cpp:
```diff
@@ -325,11 +306,12 @@ void
 		(vmThread->privateFlags & J9_PRIVATE_FLAGS_CONCURRENT_MARK_ACTIVE) &&
 		_extensions->isOld(dstObject)) {
-		/* Once dependent OMR changes are in, this should be changed to: concurrentPostWriteBarrierStore(vmThread->omrVMThread, dstObject); */
+		/* Once dependent OMR changes are in, this should be changed to: concurrentPreWriteBarrierBatchStore(vmThread->omrVMThread, dstObject); */
```
Isn't it vice versa: MM_StandardAccessBarrier::preBatchObjectStoreImpl should be renamed to be _post_?
...in which case J9ConcurrentWriteBarrierBatchStore can be replaced with concurrentPostWriteBarrierStore now.

RSalman replied:
Right, we discussed this before, was confused looking at it

RSalman replied:
Renamed `preBatchObjectStore`/`preBatchObjectStoreImpl` -> `postBatchObjectStore`/`postBatchObjectStoreImpl`

amicic commented on runtime/gc_modron_standard/StandardAccessBarrier.cpp:
```diff
@@ -933,3 +915,164 @@ MM_StandardAccessBarrier::preObjectRead(J9VMThread *vmThread, J9Object *srcObjec
+
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
+bool
+MM_StandardSATBAccessBarrier::checkClassLive(J9JavaVM *javaVM, J9Class *classPtr)
```
bad class name: MM_StandardSATBAccessBarrier

amicic commented on runtime/gc_modron_standard/StandardAccessBarrier.cpp:
```diff
@@ -933,3 +915,164 @@ MM_StandardAccessBarrier::preObjectRead(J9VMThread *vmThread, J9Object *srcObjec
+
+	UDATA offset = J9VMJAVALANGREFREFERENCE_REFERENT_OFFSET(vmThread);
+	J9Object *referent = mixedObjectReadObject(vmThread, refObject, offset, false);
+
```
part up to this point can probably be simplified, since base referenceGet, is doing effectively the same as mixedObjectReadObject

RSalman replied:
I can't seem to find the define for `J9VMJAVALANGREFREFERENCE_REFERENT_OFFSET`

RSalman replied:
The following are of interest for this:
```C++
#define J9VMJAVALANGREFREFERENCE_REFERENT_VM(javaVM, object) ((void)0, J9OBJECT_OBJECT_LOAD_VM(javaVM, object, J9VMJAVALANGREFREFERENCE_REFERENT_OFFSET_VM(javaVM)))
```
```C++
#if defined (J9VM_GC_ALWAYS_CALL_OBJECT_ACCESS_BARRIER)
#define J9OBJECT_OBJECT_LOAD_VM(javaVM, object, offset) ((j9object_t)javaVM->memoryManagerFunctions->j9gc_objaccess_mixedObjectReadObject(J9JAVAVM_VMTHREAD(javaVM), (j9object_t)(object), (offset), 0))
...
#else /* J9VM_GC_ALWAYS_CALL_OBJECT_ACCESS_BARRIER */
...
#define J9OBJECT_OBJECT_LOAD_VM(javaVM, object, offset) \
	(J9OBJECT__PRE_OBJECT_LOAD_VM(javaVM, object, offset), \
	(J9JAVAVM_COMPRESS_OBJECT_REFERENCES(javaVM) \
		? (j9object_t)J9_CONVERT_POINTER_FROM_TOKEN_VM__(javaVM, *(U_32*)((U_8*)(object) + (offset))) \
		: (j9object_t)*(UDATA*)((U_8*)(object) + (offset))))
```

RSalman replied:
Not sure if there are any performance implications to making this change. Currently, we take path 
`referenceGet` -> `J9VMJAVALANGREFREFERENCE_REFERENT_VM` which ends up doing:
```
	..... ? (j9object_t)J9_CONVERT_POINTER_FROM_TOKEN_VM__(javaVM, *(U_32*)((U_8*)(object) + (offset))) \
		: (j9object_t)*(UDATA*)((U_8*)(object) + (offset))))
```

With this change we have to go through a few different barrier methods,  `mixedObjectReadObject` -> `readObjectImpl` and some other barrier methods like `postObjectRead` and `preObjectRead` + `protectIfVolatileBefore/After`

```C+++
J9Object *
MM_ObjectAccessBarrier::mixedObjectReadObject(J9VMThread *vmThread, J9Object *srcObject, UDATA srcOffset, bool isVolatile)
{
	fj9object_t *actualAddress = J9OAB_MIXEDOBJECT_EA(srcObject, srcOffset, fj9object_t);
	J9Object *result = NULL;

	if (preObjectRead(vmThread, srcObject, actualAddress)) {
		protectIfVolatileBefore(vmThread, isVolatile, true, false);
		result = readObjectImpl(vmThread, srcObject, actualAddress, isVolatile);
		protectIfVolatileAfter(vmThread, isVolatile, true, false);

		if (!postObjectRead(vmThread, srcObject, actualAddress)) {
			result = NULL;
		}
	}
	
	/* This must always be called to massage the return value */
	return result;
}
```

RSalman replied:
Changes made

amicic commented on runtime/gc_modron_standard/StandardAccessBarrier.cpp:
```diff
@@ -933,3 +915,164 @@ MM_StandardAccessBarrier::preObjectRead(J9VMThread *vmThread, J9Object *srcObjec
+	J9Object *referent = mixedObjectReadObject(vmThread, refObject, offset, false);
+
+	/* Do nothing exceptional for NULL or marked referents */
+	if ((NULL == referent) || (globalCollector->getMarkingScheme()->isMarked(referent))) {
```
let's cache _markingScheme in this class

amicic commented on runtime/gc_modron_standard/StandardAccessBarrier.cpp:
```diff
@@ -325,11 +306,12 @@ void
 		_extensions->isOld(dstObject)) {
-		/* Once dependent OMR changes are in, this should be changed to: concurrentPostWriteBarrierStore(vmThread->omrVMThread, dstObject); */
+		/* Once dependent OMR changes are in, this should be changed to: concurrentPreWriteBarrierBatchStore(vmThread->omrVMThread, dstObject); */
 		J9ConcurrentWriteBarrierBatchStore(vmThread->omrVMThread, dstObject);
```
there is a replacement for this one too: concurrentPostWriteBarrierBatchStore

RSalman replied:
This has been replaced with: `concurrentPostWriteBarrierBatchStore`

amicic commented on runtime/gc_base/GenerationalAccessBarrierComponent.hpp:
```diff
@@ -53,7 +53,7 @@ class MM_GenerationalAccessBarrierComponent : public MM_BaseNonVirtual
 
 	void postObjectStore(J9VMThread *vmThread, J9Object *dstObject, J9Object *srcObject);
-	void preBatchObjectStore(J9VMThread *vmThread, J9Object *dstObject);
+	void postBatchObjectStore(J9VMThread *vmThread, J9Object *dstObject);
```
I believe (although not sure) that generational barrier could be either post or pre barrier (in which case these two should not have any pre or post prefix).
We should come to it eventually when doing generational+SATB, when we should try concurrent and generational together in one pre barrier call (rather than having both pre barrier called just for concurrent part and post barrier called for generational part).

RSalman replied:
Removed prefix altogether from them - makes sense not to have it.

RSalman commented:
Updated PR with latest Commit:

- All name changes in Generational Barrier reverted and the following name change made: `MM_GenerationalAccessBarrierComponent::preObjectStore` -> `MM_GenerationalAccessBarrierComponent::postObjectStore`
- `referenceGet`, `referenceReprocess`, `forcedToFinalizableObject` & `checkClassLive` refactored and simplified

amicic commented:
Jenkins test sanity all jdk11

RSalman commented:
@amicic could you please restart the tests

amicic commented:
Jenkins test sanity all jdk11

amicic commented:
Jenkins compile win32 jdk8
